plugins/PrivacyManager/Model/DataSubjects.php

```diff
@@ -0,0 +1,395 @@
+         *         $result['myplugin'] = $numDeletes;
+         *     }
+         *
+         * @param array &$results An array storing the result of how much data was deleted for .
+         * @param array &$visits An array with multiple visit entries containing an idvisit and idsite each. The data
+         *                       for these visits is requested to be deleted.
+         */
+        Piwik::postEvent('PrivacyManager.deleteDataSubjects', array(&$results, $visits));
+
+        foreach ($logTables as $logTable) {
+            $logTableName = $logTable->getName();
+            if ($logTableName === 'log_action') {
+                continue; // will be deleted via cron job automatically if the action is no longer in use
+            }
+
+            $from = array($logTableName);
+            $tableToSelect = $this->findNeededTables($logTable, $from);
+
+            if (!$tableToSelect) {
+                throw new \Exception('Cannot join table ' . $logTable->getName());
+            }
+
+            list($where, $bind) = $this->visitsToWhereAndBind($tableToSelect, $visits);
+
+            $sql = "DELETE $logTableName FROM " . $this->makeFromStatement($from) . " WHERE $where";
+
+            $result = Db::query($sql, $bind)->rowCount();
+
+            $results[$logTableName] = $result;
+        }
+
+        krsort($results); // make sure test results are always in same order
+
+        return $results;
+    }
+
+    /**
+     * @param LogTable[] $logTables
+     * @return LogTable[]
+     */
+    private function sortLogTablesToEnsureDataErasureFromAllTablesIsPossible($logTables)
+    {
+        // we need to make sure to delete eg first entries from a table log_form_field before log_form...
+        // otherwise when deleting log_form entries first, the log_form_field entries that belong to a requested visit
+        // cannot be deleted anymore as the links would be all gone. We need to make sure to delete from log_visit last
+        // and log_link_visit_action second to last.
+        usort($logTables, function ($a, $b) {
+            /** @var LogTable $a */
+            /** @var LogTable $b */
+            $aName = $a->getName();
+            $bName = $b->getName();
+            if ($bName === 'log_visit') {
+                return -1;
+            } else if ($aName === 'log_visit') {
+                return 1;
+            } else if ($bName === 'log_link_visit_action') {
+                return -1;
+            } else if ($aName === 'log_link_visit_action') {
+                return 1;
+            }
+
+            $aWays = $a->getWaysToJoinTable();
+            foreach ($aWays as $table => $column) {
+                if ($table === $bName) {
+                    return -1;
+                }
+            }
+
+            $bWays = $b->getWaysToJoinTable();
+            foreach ($bWays as $table => $column) {
+                if ($table === $aName) {
+                    return 1;
+                }
+            }
```

Comment:
Thought of a way to explain this w/ code:

instead of just `getWaysToJoinTable()`, there could be two methods like so:

```
public function getWaysToJoinTable()
{
    return [
        'log_action' => ...,
        'form_whatever' => ...,
    ];
}

public function getTablesToDeleteFromBeforeThis()
{
    return [
        'form_whatever',
    ];
}
```

Then it's way more explicit what that sort method needs.

Don't want to start an argument, though, so no worries if you don't see this.

Reply:
Not quite sure what `getTablesToDeleteFromBeforeThis ` would do? Basically, we want to always delete and export data from all log tables. It basically shouldn't be possible to say some tables cannot be deleted. But maybe I'm not quite getting it?

Reply:
> but that doesn't necessarily mean the same thing as "get tables that we should delete from before this table"

It pretty much does that mean. By the new privacy law we need to basically ensure to delete the data from all log tables, and to export any stored data.

Reply:
I mean, I imagine you could put a table in `getWaysToJoinTable()` that doesn't need to be deleted before the specific log table. Like, you can join w/ a dependent table but you can also join a table that just references the table.

Eg, let's say I have a delivery service that tracks how their drivers deliver stuff. Here a "visit" would be an entire delivery run, and it would contain multiple "actions" that are individual events (like, "product is packaged & ready to be delivered", "product is picked up by driver", "driver has not driven in ~10 mins", "driver delivered product"). If we were deleting data for a specific visit, then right now, it's straightfoward, we delete data for the "actions", then the delivery run (ie, "visit").

But, let's say the company includes other information as a visit dimension, like the car the delivery service uses. The list of possible cars are stored in a different table, like sites. This table could appear in getWaysToJoinTable(). We can join this table to get more information for a delivery run, but if we were deleting data for the run, we wouldn't delete the entry in the cars table. But from how this method is used, we would expect it to be deleted too.

This is probably not the best example, but it's the best I can come up with. Maybe getWaysToJoinTable() is only for specific types of tables that can be joined? Eg, would you put the 'site' table in there for 'log_visit'?

This discussion probably shouldn't go on too much longer though, so no worries if it still doesn't make sense.

Reply:
I think I see what you mean. It's probably more a naming issue than having to implement any changes. Here it is really meant on how to join it with other log tables (for example in order to be able to join with a visit table).

When a specific column references a site table or something, then this information would be looked up from a `Dimension` object in the `getDbColumnJoin()` method see eg https://github.com/matomo-org/matomo/blob/3.4.0/plugins/Actions/Columns/PageUrl.php#L31,  and resolve it that way without deleting anything.

Reply:
Then, would `getWaysToJoinToOtherLogTables()` or `getWaysToJoinToLogTables()` make it more clear or more complicated?

Reply:
Yes 👍